src/command_modules/azure-cli-acs/azure/cli/command_modules/acs/custom.py

```diff
@@ -753,6 +707,44 @@ def _create(resource_group_name, deployment_name, dns_name_prefix, name, ssh_key
     raise retry_exception
 
 
+def store_acs_service_principal(subscription_id, client_secret, service_principal,
+                                config_path=os.path.join(get_config_dir(),
+                                                         'acsServicePrincipal.json')):
+    obj = {}
+    if client_secret:
+        obj['client_secret'] = client_secret
+    if service_principal:
+        obj['service_principal'] = service_principal
+
+    fullConfig = load_acs_service_principals(config_path=config_path)
+    if not fullConfig:
+        fullConfig = {}
+    fullConfig[subscription_id] = obj
+
+    with os.fdopen(os.open(config_path, os.O_RDWR | os.O_CREAT | os.O_TRUNC, 0o600),
+                   'w+') as spFile:
+        json.dump(fullConfig, spFile)
+
+
+def load_acs_service_principal(subscription_id, config_path=os.path.join(get_config_dir(),
+                                                                         'acsServicePrincipal.json')):
+    config = load_acs_service_principals(config_path)
+    if not config:
+        return None
+    return config.get(subscription_id)
+
+
+def load_acs_service_principals(config_path):
+    if not os.path.exists(config_path):
+        return None
+    fd = os.open(config_path, os.O_RDONLY)
+    try:
+        with os.fdopen(fd) as f:
+            return shell_safe_json_parse(f.read())
+    except:  # pylint: disable=bare-except
+        return None
```

Comment:
Should `finally` close the file  to handle case there was an exception?

Reply:
I think I don't need a finally to close the handler according to [this](https://stackoverflow.com/questions/9885217/in-python-if-i-return-inside-a-with-block-will-the-file-still-close)

Reply:
I was referring to `os.open` above which doesn't use the `with os.open()` format and ["If fdopen() raises an exception, it leaves fd untouched (unclosed)."](https://docs.python.org/2/library/os.html#os.fdopen)

If `fdopen()` does raise an exception, I'm not sure if it would close `fd` due to you using the context manager format of `with os.fdopen()`

I guess you'd be okay though.

Reply:
Ah, we should be ok. 

Btw, then function is not introduced by this PR. I simply move it around.